### PR TITLE
Add a page for tuning KSQL in production

### PR DESCRIPTION
**Pull request by apurvam**
Need to update the `docs/installation/server-config/index.rst` file to refer to the new tuning page. Will do that once #955 goes in.

### CODE REVIEW

File: docs/installation/server-config/tuning-ksql.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
+.. _tuning_ksql:
+
+When deploying KSQL to production, we recommend adding the following settings to your ``/etc/ksql/ksql-server.properties`` file:
```
Comment (joel-hamill) [inline docs/installation/server-config/tuning-ksql.rst]: should avoid using `we` and instead write as: `When deploying KSQL to production, the following settings are recommended in your ``/etc/ksql/ksql-server.properties`` file:`

Comment (apurvam) [reply docs/installation/server-config/tuning-ksql.rst]: Thanks, updated.

File: docs/installation/server-config/index.rst [hunk 2]
```diff
@@ -161,6 +161,24 @@ listeners
...
+-----------------------------
+Production Settings
+-----------------------------
+
+When deploying KSQL to production, the following settings are recommended in your ``/etc/ksql/ksql-server.properties`` file:
+
+.. code:: bash
+
+    # Ensures that transient failures will not result in data loss.
+    producer.retries=2147483647
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: Nit: Can we clarify what this magic number represents?

Think:
```shell
# Set retries to maximum value (Integer.MAX_VALUE) to ensure that
# transient failures will not result in data loss.
producer.retries=2147483647
```

File: docs/installation/server-config/index.rst [hunk 3]
```diff
@@ -161,6 +161,24 @@ listeners
...
+
+    # Ensures that queries will not terminate if the underlying Kafka cluster is unavailable for a period of time.
+    producer.confluent.batch.expiry.ms=9223372036854775807
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: Magic number explanation like above.

```shell
# Set batch expiration to maximum value (Long.MAX_VALUE) to ensure that queries
# will not terminate if the underlying Kafka cluster is unavailable for a period of time.
producer.confluent.batch.expiry.ms=9223372036854775807
```

File: docs/installation/server-config/index.rst [hunk 4]
```diff
@@ -161,6 +161,24 @@ listeners
...
+
+    # Allows more frequent retries of requests when there are failures, enabling quicker recovery.
+    producer.request.timeout.ms=300000
 
+    # If allows KSQL to pause processing if the underlying kafka cluster is unavailable.
+    producer.max.block.ms=9223372036854775807
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: Same here (magic number).

**Comment by dguy** [comment]
@apurvam we also want to probably include:
`ksql.streams.replication.factor=3` - used when creating changelog and repartition topics. Otherwise will be set to 1
`ksql.streams.state.dir=/path/to/some/storage` - default is /tmp/kafka-streams
`ksql.streams.num.standby.replicas=2` - default is one which means that there are no replicas and if there is a node failure state will need to be rebuilt from scratch. Thus significantly slowing down the rebalance
`ksql.sink.replicas=3` (this might be `ksql.server.sink.replicas` now?) the default is 1

**Comment by miguno** [comment]
@dguy:

> `ksql.streams.num.standby.replicas=2` - default is one which means that there are no replicas and if there is a node failure state will need to be rebuilt from scratch

Wait, the default of `ksql.streams.num.standby.replicas` in KStreams is `0`, and if you want one standby replica you set the config to `1`, not to `2`?

**Comment by apurvam** [comment]
Thanks for those suggestions @dguy . Some of those were in the cloud configuration page, and some were not there at all. I incorporated all your suggestions.

**Comment by dguy** [comment]
@miguno - yes you are correct. Sorry

File: docs/installation/server-config/index.rst [hunk 5]
```diff
@@ -161,6 +161,44 @@ listeners
...
+    # Set the replication factor for internal topics, the command topic, and
+    # output topics to be 3 for better fault tolerance and durability.
+    ksql.streams.replication.factor=3
+    ksql.sink.replicas=3
+
+    # Set the storage directory for stateful operations like aggregations and
+    # joins to be at a durable location. By default, they are stored in /tmp.
+    ksql.streams.state.dir=/some/non-temporary-storage-path/
+    
+    # Bump the number of replicas for state storage for stateful operations
```
Comment (dguy) [inline docs/installation/server-config/index.rst]: @apurvam - sorry as per comment from @miguno, the default in streams is actually 0. So this should be set to >=1 to have 1 or more standby replicas

Comment (apurvam) [reply docs/installation/server-config/index.rst]: I updated the recommendation.

File: docs/installation/server-config/index.rst [hunk 6]
```diff
@@ -161,6 +161,44 @@ listeners
...
+    # time.  
+    producer.confluent.batch.expiry.ms=9223372036854775807
+
+    # Allows more frequent retries of requests when there are failures,
+    # enabling quicker recovery.
+    producer.request.timeout.ms=300000
+
+    # Set the maximum allowable time for the producer to block to
+    # Long.MAX_VALUE. This allows KSQL to pause processing if the underlying
+    # kafka cluster is unavailable.
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: Nit: "kafka cluster" -> "Kafka cluster"

File: docs/installation/server-config/index.rst [hunk 7]
```diff
@@ -161,6 +161,44 @@ listeners
...
+    # enabling quicker recovery.
+    producer.request.timeout.ms=300000
+
+    # Set the maximum allowable time for the producer to block to
+    # Long.MAX_VALUE. This allows KSQL to pause processing if the underlying
+    # kafka cluster is unavailable.  
+    producer.max.block.ms=9223372036854775807
+
+    # Set the replication factor for internal topics, the command topic, and
+    # output topics to be 3 for better fault tolerance and durability.
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: We should add sth like:
> Note: The value 3 requires that there are at least 3 brokers in your Kafka cluster.

**Comment by miguno** [comment]
@joel-hamill : Could you please follow-up here?

Where to put this content: My first suggestion would be to make it a sub-section of "Configuring KSQL Server".

File: docs/installation/server-config/index.rst [hunk 8]
```diff
@@ -184,11 +184,12 @@ When deploying KSQL to production, the following settings are recommended in you
...
     # Set the maximum allowable time for the producer to block to
     # Long.MAX_VALUE. This allows KSQL to pause processing if the underlying
-    # kafka cluster is unavailable.  
+    # Kafka cluster is unavailable.  
     producer.max.block.ms=9223372036854775807
 
     # Set the replication factor for internal topics, the command topic, and
-    # output topics to be 3 for better fault tolerance and durability.
+    # output topics to be 3 for better fault tolerance and durability. Note:
+    # the value 3 requires at least 3 brokers in your kafka cluster.
```
Comment (miguno) [inline docs/installation/server-config/index.rst]: Nit: "kafka cluster" -> "Kafka cluster".

**Pull request merged**